node/src/blockchain/blockchain_interface/data_structures/mod.rs

```diff
@@ -37,11 +44,184 @@ pub struct RetrievedBlockchainTransactions {
 pub struct RpcPayableFailure {
     pub rpc_error: Error,
     pub recipient_wallet: Wallet,
-    pub hash: H256,
+    pub hash: TxHash,
 }
 
 #[derive(Debug, PartialEq, Clone)]
 pub enum ProcessedPayableFallible {
     Correct(PendingPayable),
     Failed(RpcPayableFailure),
 }
+
+#[derive(Debug, PartialEq, Eq, Message, Clone)]
+pub struct TxReceiptResult(pub Result<RetrievedTxStatus, TxReceiptError>);
+
+impl TxReceiptResult {
+    pub fn hash(&self) -> TxHashByTable {
+        match &self.0 {
+            Ok(retrieved_tx_status) => retrieved_tx_status.tx_hash,
+            Err(tx_receipt_error) => tx_receipt_error.tx_hash,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct RetrievedTxStatus {
+    pub tx_hash: TxHashByTable,
+    pub status: StatusReadFromReceiptCheck,
+}
+
+impl RetrievedTxStatus {
+    pub fn new(tx_hash: TxHashByTable, status: StatusReadFromReceiptCheck) -> Self {
+        Self { tx_hash, status }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum StatusReadFromReceiptCheck {
+    Failed(BlockchainTxFailure),
+    Succeeded(TxBlock),
+    Pending,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, VariantCount)]
+pub enum BlockchainTxFailure {
+    Unrecognized,
+}
+
+impl Display for BlockchainTxFailure {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            BlockchainTxFailure::Unrecognized => write!(f, "Unrecognized failure"),
+        }
+    }
+}
+
+impl Display for StatusReadFromReceiptCheck {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StatusReadFromReceiptCheck::Failed(reason) => {
+                write!(f, "Failed(Reason: {})", reason)
+            }
+            StatusReadFromReceiptCheck::Succeeded(block) => {
+                write!(
+                    f,
+                    "Succeeded({},{:?})",
+                    block.block_number, block.block_hash
+                )
+            }
+            StatusReadFromReceiptCheck::Pending => write!(f, "Pending"),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct TxReceiptError {
+    pub tx_hash: TxHashByTable,
+    pub err: AppRpcError,
+}
+
+impl TxReceiptError {
+    pub fn new(tx_hash: TxHashByTable, err: AppRpcError) -> Self {
+        Self { tx_hash, err }
+    }
+}
+
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy, Ord, PartialOrd, Serialize, Deserialize)]
+pub struct TxBlock {
```

Comment:
I have noticed that the fields of this structure can also be found in the `TxStatus`. I wonder if it's a common concept then we should use this one everywhere, and I'd appreciate if you prefer not to use `TxHash` and `u64` data types here, that'll eliminate the dependency on web3 code.

Reply:
It cannot be done as `H256` is too big a number for `Serde` and thus I use it in the String form in `TxStatus`. The different representation of the block hash causes us to have two data types instead of using just one. 

Otherwise you're right.